### DISCUSSION TITLE
Custom SQL queries against internal table?

### DISCUSSION THREAD

**Discussion by chrisshaw** [Budibase/budibase#1939] [Help] [answered]
While reading https://github.com/Budibase/budibase/discussions/1811 I noticed a quick remark by @mjashanks that you can write custom SQL queries. Is that against **BB's internal tables?** I've run into this need once or twice around nested object models as a sane alternative to nested Data Providers. (Unless I'm missing some other obvious pattern.)

> May I also say kudos to this small team and their responsiveness to the community? :bow:

**Comment by mike12345567** [comment, marked as answer]
Hi @chrisshaw - sadly this isn't something that is currently supported, the underlying database of Budibase is actually CouchDB, a NoSQL document store, we use the internal database for everything that makes up an app, as well as everything that makes the Budibase platform work (as well the data stored in the internal DB). Sadly however this does mean it can't perform complex querying capabilities (like wide row responses with deep joining in SQL).

The next release of Budibase will include an even deeper integration with SQL databases, specifically allowing nearly all the functionality you can get with the internal DB against SQL databases, which also comes with the benefit of allowing creation of very complex queries if the base functionality provided by Budibase isn't enough. You can see the deeper integration with SQL [here](https://github.com/Budibase/budibase/pull/1907) (specifically providing SQL relationship integration).

**Comment by shogunpurple** [comment]
Hi @chrisshaw, first of all thank you for the kind words! We appreciate you taking the time and effort to help us improve budibase.

Just for additional context: The custom SQL queries are against Postgres, MySQL and MSSQL databases. Here's the documentation on this feature:
https://docs.budibase.com/data/data-sources

This is more suited for very complex or custom queries. As @mike12345567 mentioned, for almost all CRUD based workflows you should be able to use our opinionated datasource feature.

**Reply by chrisshaw** [reply]
Thanks @shogunpurple , just as a gut-check: for that nested data issue I mentioned, is the "right way" to do that through nested data providers and filters?

For context:

![image](https://user-images.githubusercontent.com/2966918/124624939-5bdc2680-de4b-11eb-8650-7a8e58fdaeea.png)

**Reply by shogunpurple** [reply]
Yep, that's correct, you essentially render a data provider for every nested level of data. Again this is something at the forefront of our thinking - we are planning to add better mapping/transforming capabilities to the builder to make these use cases easier.
